module/zfs/zfs_dir.c

```diff
@@ -508,8 +510,55 @@ zfs_unlinked_drain(zfsvfs_t *zfsvfs)
 
 		zp->z_unlinked = B_TRUE;
 		iput(ZTOI(zp));
+		ASSERT(!zfsvfs->z_unmounted);
 	}
 	zap_cursor_fini(&zc);
+
+	mutex_enter(&zfsvfs->z_drain_lock);
+	zfsvfs->z_drain_state = ZFS_DRAIN_SHUTDOWN;
+	cv_broadcast(&zfsvfs->z_drain_cv);
+	mutex_exit(&zfsvfs->z_drain_lock);
+}
+
+/*
+ * Sets z_drain_state then tries to dispatch async unlinked drain.
+ * If that fails executes synchronous unlinked drain.
+ */
+void
+zfs_unlinked_drain(zfsvfs_t *zfsvfs)
+{
+	ASSERT(!zfsvfs->z_unmounted);
+
+	mutex_enter(&zfsvfs->z_drain_lock);
+	ASSERT3U(zfsvfs->z_drain_state, ==, ZFS_DRAIN_SHUTDOWN);
+	zfsvfs->z_drain_state = ZFS_DRAIN_RUNNING;
+	zfsvfs->z_drain_continue = B_TRUE;
+	mutex_exit(&zfsvfs->z_drain_lock);
+
+	if (taskq_dispatch(dsl_pool_unlinked_drain_taskq(
+	    dmu_objset_pool(zfsvfs->z_os)),
+	    (void (*)(void *))zfs_unlinked_drain_task, zfsvfs,
+	    TQ_NOSLEEP) == TASKQID_INVALID) {
```

Comment:
The problem here is that draining is not interruptable until the task is running. My thought was that if we can't get the task to run right away I would rather have it fail and do a sync unlink drain than not schedule the task, and as a result, hold the unmount of the dataset.

I guess it doesn't really matter since if we're out of memory there are going to be other more serious problems.

Reply:
My concern is that using `TQ_NOSLEEP` makes a failure more likely since it disables the direct memory reclaim call paths, where `TQ_SLEEP` will wait for the needed memory.  Though as you say it probably doesn't matter to much since we're only talking about a handful of bytes here for the taskq entry, so any failure is unlikely.  But I do think `TQ_SLEEP` is slightly better since if the kernel can't give us a handful of bytes we're in a very bad spot.